[Bug]: agent_id do not support underscore or dash chars

### Describe the bug

agent_id do not support underscore

If agent_id has underscore, the add memory fails.
See attachement a5.txt

However, group_id, user_id and session_id seem to support underscores.

The online doc example tells the user to use underscore.
https://docs.memmachine.ai/examples/rest


POST /v1/memories

```
Traceback (most recent call last):
  File "/home/ec2-user/mrcr/tom_mrcr_memmachine.py", line 414, in <module>
    stat = run_test(args.model, args.max_context, args.skip, args.count, log_fp, args.memmachine_url)
  File "/home/ec2-user/mrcr/tom_mrcr_memmachine.py", line 314, in run_test
    memory_manager.add_memory(role, content) 
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/home/ec2-user/mrcr/tom_mrcr_memmachine.py", line 104, in add_memory
    raise AssertionError(f"Failed to add episodic memory: url={self.add_url} headers={headers} payload={payload} status={resp.status_code} resp={resp.text}")
AssertionError: Failed to add episodic memory: url=http://3.143.227.240:8080/v1/memories headers={'Content-Type': 'application/json', 'Accept': 'application/json, text/javascript, */*', 'X-Requested-With': 'XMLHttpRequest', 'Sec-Fetch-Site': 'same-origin', 'Sec-Fetch-Mode': 'cors', 'Sec-Fetch-Dest': 'empty', 'Accept-Encoding': 'gzip, deflate, br'} payload={'session': {'group_id': 'test_group', 'agent_id': ['test_agent'], 'user_id': ['user', 'assistant'], 'session_id': 'session_3103224'}, 'producer': 'user', 'produced_for': 'test_agent', 'episode_content': 'write a song about hearts', 'episode_type': 'message'} status=500 resp=Internal Server Error
```



### Steps to reproduce

POST /v1/memories
where payload={'session': {'group_id': 'test_group', 'agent_id': ['test_agent'], 'user_id': ['user', 'assistant'], 'session_id': 'session_3103224'}, 'producer': 'user', 'produced_for': 'test_agent', 'episode_content': 'write a song about hearts', 'episode_type': 'message'}


### Expected behavior

Following the example from the online doc should work, but it doesn't.
https://docs.memmachine.ai/examples/rest


### Environment

amazon linux 2023
I am using main/HEAD. I did a git clone on 10/24 17:21. commit e39d45b8bbe3ed1e1d3d8b2bb9efee66a1bc1707


### Additional context

[a5.txt](https://github.com/user-attachments/files/23197817/a5.txt)

--- a/src/memmachine/server/app.py
+++ b/src/memmachine/server/app.py
@@ -22,7 +22,7 @@ from typing import Any, Self, cast
 import uvicorn
 import yaml
 from dotenv import load_dotenv
-from fastapi import FastAPI, Header, HTTPException
+from fastapi import FastAPI, Header, HTTPException, Request
 from fastapi.exceptions import RequestValidationError
 from fastapi.params import Depends
 from fastapi.responses import Response
@@ -155,7 +155,7 @@ class SessionData(BaseModel):
     )
 
     user_id: list[str] = Field(
-        default=[AppConst.DEFAULT_USER_ID],
+        default=[],
         description=AppConst.USER_ID_DOC,
         examples=AppConst.USER_ID_EXAMPLES,
     )
@@ -180,11 +180,14 @@ class SessionData(BaseModel):
                 ret = a or b
             return sorted(ret)
 
-        if other.group_id:
+        if other.group_id and other.group_id != AppConst.DEFAULT_GROUP_ID:
             self.group_id = other.group_id
 
-        if other.session_id:
+        if other.session_id and other.session_id != AppConst.DEFAULT_SESSION_ID:
             self.session_id = other.session_id
+
+        if other.user_id == [AppConst.DEFAULT_USER_ID]:
+            other.user_id = []
 
         self.agent_id = merge_lists(self.agent_id, other.agent_id)
         self.user_id = merge_lists(self.user_id, other.user_id)
@@ -224,7 +227,7 @@ class SessionData(BaseModel):
     @model_validator(mode="after")
     def _set_default_user_id(self) -> Self:
         """Defaults user_id to ['default'] if not set."""
-        if len(self.user_id) == 0:
+        if len(self.user_id) == 0 and len(self.agent_id) == 0:
             self.user_id = [AppConst.DEFAULT_USER_ID]
         else:
             self.user_id = sorted(self.user_id)
@@ -391,6 +394,7 @@ def _split_str_to_list(s: str) -> list[str]:
 
 
 async def _get_session_from_header(
+    request: Request,
     group_id: str = Header(
         AppConst.DEFAULT_GROUP_ID,
         alias=AppConst.GROUP_ID_KEY,
@@ -417,6 +421,23 @@ async def _get_session_from_header(
     ),
 ) -> SessionData:
     """Extract session data from headers and return a SessionData object."""
+    group_id_keys = [AppConst.GROUP_ID_KEY, "group_id"]
+    session_id_keys = [AppConst.SESSION_ID_KEY, "session_id"]
+    agent_id_keys = [AppConst.AGENT_ID_KEY, "agent_id"]
+    user_id_keys = [AppConst.USER_ID_KEY, "user_id"]
+    headers = request.headers
+
+    def get_with_alias(possible_keys: list[str], default: str):
+        for key in possible_keys:
+            for hk, hv in headers.items():
+                if hk.lower() == key.lower():
+                    return hv
+        return default
+
+    group_id = get_with_alias(group_id_keys, group_id)
+    session_id = get_with_alias(session_id_keys, session_id)
+    agent_id = get_with_alias(agent_id_keys, agent_id)
+    user_id = get_with_alias(user_id_keys, user_id)
     return SessionData(
         group_id=group_id,
         session_id=session_id,

--- a/tests/memmachine/server/test_request_parameter.py
+++ b/tests/memmachine/server/test_request_parameter.py
@@ -66,3 +66,46 @@ def test_default_producer_with_empty_session():
     session = SessionData()
     episode = NewEpisode(session=session)
     assert episode.producer == "default"
+
+
+def test_merge_session_data_with_default():
+    session = SessionData(
+        user_id=["dave", "carol"],
+        group_id="group_a",
+        session_id="session_b",
+        agent_id=["my_agent", "another_agent"],
+    )
+    dft_session = SessionData()
+    session.merge(dft_session)
+    assert session.user_id == ["carol", "dave"]
+    assert session.group_id == "group_a"
+    assert session.session_id == "session_b"
+    assert session.agent_id == ["another_agent", "my_agent"]
+
+
+def test_no_default_agent_if_user_exists():
+    session = SessionData(
+        user_id=["dave", "carol"],
+        group_id="group_a",
+        session_id="session_b",
+    )
+    dft_session = SessionData()
+    session.merge(dft_session)
+    assert session.user_id == ["carol", "dave"]
+    assert session.group_id == "group_a"
+    assert session.session_id == "session_b"
+    assert session.agent_id == []
+
+
+def test_no_default_user_if_agent_exists():
+    session = SessionData(
+        group_id="group_a",
+        session_id="session_b",
+        agent_id=["my_agent", "another_agent"],
+    )
+    dft_session = SessionData()
+    session.merge(dft_session)
+    assert session.user_id == []
+    assert session.group_id == "group_a"
+    assert session.session_id == "session_b"
+    assert session.agent_id == ["another_agent", "my_agent"]

--- a/tests/memmachine/server/test_rest_memories.py
+++ b/tests/memmachine/server/test_rest_memories.py
@@ -144,10 +144,10 @@ def valid_session_headers():
 def alias_session_headers():
     """Provides valid headers using alias names for session fields."""
     return {
-        "group-id": "group-alias",
-        "session-id": "session-alias",
-        "agent-id": "agent5,agent6",
-        "user-id": "user5,user6",
+        "group_id": "group-alias",
+        "session_id": "session-alias",
+        "agent_id": "agent5,agent6",
+        "user_id": "user5,user6",
     }
 
 

